gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystem.java

```diff
@@ -404,10 +414,59 @@ public void delete(URI path, boolean recursive) throws IOException {
               : listFileInfoForPrefixPage(
                       fileInfo.getPath(), DELETE_RENAME_LIST_OPTIONS, /* pageToken= */ null)
                   .getItems();
-      if (!itemsToDelete.isEmpty() && !recursive) {
+
+      if (isHnBucket) {
+        /**
+         * Get list of folders in path if the bucket is HN enabled bucket For recursive delete, get
+         * all folder resources. For non-recursive delete, only get maximum of 1 folder resources
+         * other than the path.
+         */
+        listOfFolders =
```

Comment:
why not do this after the objects are deleted? Now it will have to keep track of both the objects and folders in memory, correct?

Reply:
Will add a todo for this

Reply:
let us go with current flow of code and later we can decide to optimise it. Added a TODO. 
my main concern is non-recursive delete, since we need to check that both objects and folder resources must be empty.